File: Python3/3.Longest_Substring_Without_Repeating.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) <= 1:
            return len(s)
        total = 1
        i = j = 0
        contains = dict()
        for j in range(len(s)):
            if s[j] in contains.keys():
                i = max(contains[s[j]], i)
            total = max(j - i + 1, total)
            contains[s[j]] = j + 1
        return total

        '''
        contains.add(s[0])

        while j < len(s) - 1:
            j += 1
            if s[j] not in contains:
                contains.add(s[j])
                total = max(j - i + 1, total)
            else:
                while 1:
                    if i == j:
                        break
                    elif s[i] == s[j]:
                        i += 1
                        break
                    else:
                        contains.remove(s[i])
                        i += 1
        '''
        return total

File: Java/3.Longest_Substring_Without_Repeating.java
import java.util.HashMap;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();
        int total = 0;

        for(int i = 0 , j = 0 ; j < s.length(); ++j) {
            if(map.containsKey(s.charAt(j))) {
                i = Math.max(i, map.get(s.charAt(j)));
            }
            total = Math.max(j - i + 1, total);
            map.put(s.charAt(j), j + 1);
        }
        return total;

    }
}
